app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,166 @@
+
+      reason = if topic && !user.guardian.can_see?(topic)
+        topic.private_message? ? :private : :category
+      elsif allowed_names.present? && !is_user_allowed?(user, topic_allowed_user_ids, topic_allowed_group_ids)
+        # This would normally be handled by the previous if, but that does not work for new private messages.
+        :private
+      elsif topic_muted_by.include?(user.id)
+        :muted_topic
+      elsif topic&.private_message? && !is_user_allowed?(user, topic_allowed_user_ids, topic_allowed_group_ids)
+        # Admins can see the topic, but they will not be mentioned if they were not invited.
+        :not_allowed
+      end
+
+      # Regular users can see only basic information why the users cannot see the topic.
+      reason = nil if !guardian.is_staff? && reason != :private && reason != :category
+
+      user_reasons[username] = reason if reason.present?
+    end
+
+    names.each do |name|
+      group = groups[name]
+      next if group.blank?
+
+      group_reasons[name] = if !mentionable_group_ids.include?(group.id)
+        :not_mentionable
+      elsif (topic&.private_message? || allowed_names.present?) && !topic_allowed_group_ids.include?(group.id)
+        :not_allowed
+      end
+    end
+
+    if topic && names.include?(SiteSetting.here_mention) && guardian.can_mention_here?
+      here_count = PostAlerter.new.expand_here_mention(topic.first_post, exclude_ids: [current_user.id]).size
+    end
+
+    serialized_groups = groups.values.map do |group|
+      serialized_group = { user_count: group.user_count }
+
+      if group_reasons[group.name] == :not_allowed &&
+          members_visible_group_ids.include?(group.id) &&
+          (topic&.private_message? || allowed_names.present?)
+
+        # Find users that are notified already because they have been invited
+        # directly or via a group
+        notified_count = GroupUser
+          # invited directly
+          .where(user_id: topic_allowed_user_ids)
+          .or(
+            # invited via a group
+            GroupUser.where(
+              user_id: GroupUser.where(group_id: topic_allowed_group_ids).select(:user_id)
+            )
+          )
+          .where(group_id: group.id)
+          .select(:user_id).distinct.count
+
+        if notified_count > 0
+          group_reasons[group.name] = :some_not_allowed
+          serialized_group[:notified_count] = notified_count
+        end
+      end
+
+      [group.name, serialized_group]
+    end
+
+    render json: {
+      users: users.keys,
+      user_reasons: user_reasons.compact,
+      groups: serialized_groups.to_h,
+      group_reasons: group_reasons.compact,
+      here_count: here_count,
+      max_users_notified_per_group_mention: SiteSetting.max_users_notified_per_group_mention,
+    }
+  end
+
+  private
+
+  def is_user_allowed?(user, user_ids, group_ids)
+    user_ids.include?(user.id) || user.group_ids.any? { |group_id| group_ids.include?(group_id) }
```

Comment:
🏌️ ⛳ 
(but only if you find it as readable as the current version 😉 )

```suggestion
    user_ids.include?(user.id) || (user.group_ids & group_ids).any?
```

Reply:
I just realized this does not work one because one of them is an array and the other one is a set. Anyway, I don't think it makes sense to do the complete union if we care about only if there is some overlap.